remill/BC/DeadStoreEliminator.cpp

```diff
@@ -0,0 +1,532 @@
+        return VisitResult::NoProgress;
+      } else {
+        if (!in_offset_map) {
+          offset = ptr->second;
+          in_offset_map = true;
+        } else {
+          if (ptr->second != offset) {
+            // bail if the offsets don't match
+            return VisitResult::Error;
+          }
+        }
+      }
+    }
+  }
+  if (in_offset_map) {
+    LOG(INFO) << "offsetting: " << LLVMThingToString(&I);
+    offset_map.insert({&I, offset});
+  } else {
+    LOG(INFO) << "excluding: " << LLVMThingToString(&I);
+    exclude.insert(&I);
+  }
+  return VisitResult::Progress;
+}
+
+// Return the scope of the given instruction.
+llvm::MDNode *GetScopeFromInst(llvm::Instruction &I) {
+  llvm::AAMDNodes N;
+  I.getAAMetadata(N);
+  return N.Scope;
+}
+
+// Remove all dead stores.
+void removeDeadStores(std::unordered_set<llvm::Instruction *> dead_stores) {
+  /* for (auto *store : dead_stores) { */
+  /*   store->eraseFromParent(); */
+  /* } */
+}
+
+LiveSetBlockVisitor::LiveSetBlockVisitor(std::unordered_map<llvm::MDNode *, uint64_t> scope_to_offset_)
+  : scope_to_offset(scope_to_offset_), curr_wl(), next_wl(), block_map(), to_remove(), live() { }
+
+void LiveSetBlockVisitor::addFunction(llvm::Function &func) {
+  for (auto &block : func) {
+    // The machines rose from the ashes of the nuclear fire....
+    if (block.getTerminator()) {
+      curr_wl.push_back(&block);
+      // add block to map with initial entry and exit sets
+      block_map.insert({&block, std::make_pair(LiveSet(), LiveSet())});
+    }
+  }
+}
+
+void LiveSetBlockVisitor::visit() {
+  //bool progress = true;
+  // if any visit makes progress, continue the loop
+  while (!curr_wl.empty()/* && progress*/) {
+    LOG(INFO) << "LSBV: " << curr_wl.size() << " blocks in worklist";
+    //progress = false;
+    for (auto block : curr_wl) {
+      switch (visitBlock(block)) {
+        case VisitResult::Progress:
+          //progress = true;
+          break;
+        case VisitResult::NoProgress:
+          next_wl.push_back(block);
+          break;
+        case VisitResult::Error:
+          return;
+      }
+      live = block_map[block].first;
+      LOG(INFO) << "LSBV: " << live.count() << " live slots";
+    }
+    // update curr_wl to the next set of insts
+    curr_wl.swap(next_wl);
+    // reset upcoming insts
+    next_wl.clear();
+  }
+  LOG(INFO) << "Found " << to_remove.size() << " dead stores to remove";
+  removeDeadStores(to_remove);
+}
+
+VisitResult LiveSetBlockVisitor::visitBlock(llvm::BasicBlock *B) {
+  for (llvm::BasicBlock *succ : successors(B)) {
+    // all successors must be default have been visited first
+    // b.exit = U { s.entry | for s in successors(b)}
+    block_map[B].second |= block_map[succ].first;
+  }
+  for (auto inst = B->rbegin(); inst != B->rend(); ++inst) {
+    if (inst != B->rbegin() && llvm::isa<llvm::TerminatorInst>(&*inst)) {
+      // mark as all live
+      block_map[B].first = LiveSet(4095);
```

Comment:
I am pretty sure that this doesn't do what you think it does. Also, maps/unordered maps default initialize their things. So `.first` is already a "valid" bitset, with all bits cleared. If you want to set all bits to `1`, then what you want is `block_map[B].first.set();`.